editor/osm_auth.cpp

```diff
@@ -0,0 +1,418 @@
+  return OsmOAuth(OSM_CONSUMER_KEY, OSM_CONSUMER_SECRET, kOsmMainSiteURL, kOsmApiURL);
+}
+
+OsmOAuth OsmOAuth::ServerAuth()
+{
+  // TODO(AlexZ): Replace with ProductionServerAuth before release.
+  return IZServerAuth();
+}
+
+// Opens a login page and extract a cookie and a secret token.
+OsmOAuth::AuthResult OsmOAuth::FetchSessionId(OsmOAuth::SessionID & sid) const
+{
+  HTTPClientPlatformWrapper request(m_baseUrl + "/login?cookie_test=true");
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  sid.m_cookies = request.combined_cookies();
+  sid.m_token = findAuthenticityToken(request.server_response());
+  return !sid.m_cookies.empty() && !sid.m_token.empty() ? AuthResult::OK : AuthResult::FailCookie;
+}
+
+// Log a user out.
+OsmOAuth::AuthResult OsmOAuth::LogoutUser(SessionID const & sid) const
+{
+  HTTPClientPlatformWrapper request(m_baseUrl + "/logout");
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  return AuthResult::OK;
+}
+
+// Signs a user id using login and password.
+OsmOAuth::AuthResult OsmOAuth::LoginUserPassword(string const & login, string const & password, SessionID const & sid) const
+{
+  map<string, string> params;
+  params["username"] = login;
+  params["password"] = password;
+  params["referer"] = "/";
+  params["commit"] = "Login";
+  params["authenticity_token"] = sid.m_token;
+  HTTPClientPlatformWrapper request(m_baseUrl + "/login");
+  request.set_body_data(buildPostRequest(params), "application/x-www-form-urlencoded");
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  if (!request.was_redirected())
+    return AuthResult::FailLogin;
+  // Since we don't know whether the request was redirected or not, we need to check page contents.
+  return isLoggedIn(request.server_response()) ? AuthResult::OK : AuthResult::FailLogin;
+}
+
+// Signs a user in using a facebook token.
+OsmOAuth::AuthResult OsmOAuth::LoginSocial(string const & callbackPart, string const & socialToken, SessionID const & sid) const
+{
+  string const url = m_baseUrl + callbackPart + socialToken;
+  HTTPClientPlatformWrapper request(url);
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  if (!request.was_redirected())
+    return AuthResult::FailLogin;
+  return isLoggedIn(request.server_response()) ? AuthResult::OK : AuthResult::FailLogin;
+}
+
+// Fakes a buttons press, so a user accepts requested permissions.
+string OsmOAuth::SendAuthRequest(string const & requestTokenKey, SessionID const & sid) const
```

Comment:
The whole workflow including this function is tested in `osm_auth_tests.cpp`. The function itself is private.